kvserver: test snapshot preparation

pav-kv opened the pull request:
This PR introduces a basic datadriven test for `prepareSnapApply` path.

TODO for follow-ups:

1. [ ] the excised span [can be wider](https://github.com/cockroachdb/cockroach/pull/147268#issuecomment-2908006179) when the subsumed ranges are wider than the ingested snapshot
2. [ ] seems like some range dels are written twice in [this](https://github.com/cockroachdb/cockroach/blob/c81390fdc04781b5cf2440fc9bc49d49b4265457/pkg/kv/kvserver/replica_raftstorage.go#L620-L624) path (probably benign, but can be deduped)

Epic: CRDB-46488

pav-kv committed:
kvserver: test snapshot preparation

Epic: none
Release note: none

pav-kv commented:
I think we have a bug in [this](https://github.com/cockroachdb/cockroach/blob/c81390fdc04781b5cf2440fc9bc49d49b4265457/pkg/kv/kvserver/replica_raftstorage.go#L614) line. The excised span should be wider if the subsumed ranges expand beyond the snapshotted range's end key. Seems similar to #73462.

pav-kv commented on pkg/kv/kvserver/testdata/TestPrepareSnapApply.txt:
```diff
@@ -0,0 +1,28 @@
+>> sst:
+set: /Local/RangeID/102/u/RangeTombstone/0,0 -> /<err: unknown tag: 16>
+rangedel: /Local/RangeID/102/{u"rftr"-v""}
+>> excise: {a-k}
```
bug: should be `{a-z}`

pav-kv commented on pkg/kv/kvserver/testdata/TestPrepareSnapApply.txt:
```diff
@@ -0,0 +1,28 @@
+set: /Local/RangeID/102/u/RangeTombstone/0,0 -> /<err: unknown tag: 16>
+rangedel: /Local/RangeID/102/{u"rftr"-v""}
+>> excise: {a-k}
+>> repl: /Local/RangeID/123/{r""-s""}
+>> repl: /Local/Range"{a"-k"}
+>> repl: /Local/Lock/Local/Range"{a"-k"}
+>> repl: /Local/Lock"{a"-k"}
```
There seems to be nothing that clears these spans in the [IngestAndExcise](https://github.com/cockroachdb/cockroach/blob/c81390fdc04781b5cf2440fc9bc49d49b4265457/pkg/kv/kvserver/replica_raftstorage.go#L615-L618) path? We only excise the single MVCC span, but the other Range-local ones are only cleared in the "[convert to batch](https://github.com/cockroachdb/cockroach/blob/c81390fdc04781b5cf2440fc9bc49d49b4265457/pkg/kv/kvserver/replica_raftstorage.go#L620-L621)" case.

tbg replied:
I think this is taken care of here (i.e. in code this test doesn't see):

https://github.com/cockroachdb/cockroach/blob/33b1a4690641beae63bbc2679eddd52e11d85e1f/pkg/kv/kvserver/kvstorage/snaprecv/multi_sst_writer.go#L236-L256

We do this for all but the last span in

https://github.com/cockroachdb/cockroach/blob/33b1a4690641beae63bbc2679eddd52e11d85e1f/pkg/kv/kvserver/kv_snapshot_strategy.go#L100-L102

which is backed by:

https://github.com/cockroachdb/cockroach/blob/33b1a4690641beae63bbc2679eddd52e11d85e1f/pkg/kv/kvserver/rditer/select.go#L66-L120

In other words, I think the lock table is already cleared in the SSTs we already have on file when we enter `applySnapshotRaftMuLocked`.

Since I'm here - the `rditer.Select` interface is quite messy. These "filters":

https://github.com/cockroachdb/cockroach/blob/33b1a4690641beae63bbc2679eddd52e11d85e1f/pkg/kv/kvserver/rditer/select.go#L21-L37

Seem too convoluted. I'll send a PR to clean this up.

pav-kv committed:
storageutils: use engine keys in ReportSSTEntries

STTs can contain keys other than MVCCKey, such as lock table keys.

Epic: none
Release note: none

pav-kv committed:
print: print LockTableKey

Epic: none
Release note: none

pav-kv commented on pkg/kv/kvserver/testdata/TestPrepareSnapApply.txt:
```diff
@@ -0,0 +1,32 @@
+echo
+----
+>> sst:
+set: /Local/RangeID/123/u/RaftHardState/0,0 -> term:20 vote:0 commit:100 lead:0 lead_epoch:0 
+set: /Local/RangeID/123/u/RaftReplicaID/0,0 -> replica_id:4 
+set: /Local/RangeID/123/u/RaftTruncatedState/0,0 -> index:100 term:20 
+rangedel: /Local/RangeID/123/{u""-v""}
```
I suppose this reflects the fact that point and range-dels/keys are in different blocks in SST. So this is just a raw dump. Interesting though how Pebble disambiguates `put(a)+rangedel[a,b)` vs `rangedel[a,b)+put(a)` - or is `a` not allowed to be present twice in this way?

pav-kv replied:
Left a TODO in the code to do something about it.

pav-kv replied:
Incidentally found a way to fix this order: #149499

pav-kv commented:
bors r=tbg
